### PR TITLE
Multiple llama4 fixe

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?

- Fixed training with flex
- Fixed inference with flex, no cache
- Fixed full compile compatibility 
- Reduced overhead for the cache cc @Cyrilvallez

### CODE REVIEW

File: src/transformers/models/llama4/modeling_llama4.py [hunk 1]
```diff
@@ -761,27 +761,34 @@ def _update_causal_mask(
...
+
         if self.config._attn_implementation == "flex_attention":
             if isinstance(attention_mask, torch.Tensor):
                 offsets = (first_cache_position, max(last_cache_position - key_length, 0))
                 chunked_attention_mask = make_flex_block_causal_mask(
                     attention_mask, self.config.attention_chunk_size, sequence_length, key_length, offsets=offsets
                 )
+                full_key_length = (
+                    past_key_values.get_max_cache_shape() | target_length if not self.training else target_length
+                )
```
Comment (ArthurZucker) [inline src/transformers/models/llama4/modeling_llama4.py:787]: fixes training

**Review by ArthurZucker — COMMENTED** [review]
Hope this does not break the graph for TGI

File: src/transformers/models/llama4/modeling_llama4.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/llama4/modeling_llama4.py:787]: I'm still a bit concerned about `past_key_values.get_max_cache_shape()`, e.g. if inference doesn't use a cache - but that's an edge case. 

Good luck on the fixes!

File: src/transformers/integrations/flex_attention.py [hunk 2]
```diff
@@ -206,20 +207,20 @@ def flex_attention_forward(
     causal_mask = None
     if isinstance(attention_mask, BlockMask):
         block_mask = attention_mask
+        score_mod = None
     else:
         causal_mask = attention_mask
 
-    if causal_mask is not None:
         causal_mask = causal_mask[:, :, :, : key.shape[-2]]
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:214]: This seems breaking to me as the attention mask could be `None`

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -1953,13 +1964,6 @@ def update(
         cache_position = cache_kwargs.get("cache_position")
         self.initialise_cache_layer(layer_idx, key_states)
 
-        # These two `if` blocks are only reached in multigpu and if `layer_device_map` is not passed. They are used
-        # when the cache is initialized in the forward pass (e.g. Gemma2)
-        if self.key_cache[layer_idx].device != key_states.device:
-            self.key_cache[layer_idx] = self.key_cache[layer_idx].to(key_states.device)
-        if self.value_cache[layer_idx].device != value_states.device:
-            self.value_cache[layer_idx] = self.value_cache[layer_idx].to(value_states.device)
```
Comment (vasqu) [inline src/transformers/cache_utils.py:1961]: Any particular reason this has been removed?

File: src/transformers/integrations/flex_attention.py [hunk 4]
```diff
@@ -206,20 +207,20 @@ def flex_attention_forward(
...
-            score = score + head_mask[batch_idx][head_idx][0][0]
-        return score
+        def score_mod(score, batch_idx, head_idx, q_idx, kv_idx):
+            if softcap is not None:
+                score = softcap * torch.tanh(score / softcap)
+            if causal_mask is not None:
+                score = score + causal_mask[batch_idx][0][q_idx][kv_idx]
+            if head_mask is not None:
+                score = score + head_mask[batch_idx][head_idx][0][0]
+            return score
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:223]: This change seems suboptimal to me -> gemma2 wouldn't use softcapping anymore

**Review by vasqu — COMMENTED** [review]
Some smaller questions on my side

File: src/transformers/cache_utils.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/cache_utils.py:1961]: The layers are initialized on the correct device now!!
